src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -0,0 +1,259 @@
+#include "model/metadata.h"
+#include "model/namespace.h"
+#include "model/timeout_clock.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/sstring.hh>
+
+#include <absl/container/node_hash_map.h>
+#include <fmt/format.h>
+
+#include <algorithm>
+#include <chrono>
+#include <iterator>
+#include <system_error>
+#include <vector>
+
+namespace kafka {
+
+using partitions_request_iterator
+  = std::vector<reassignable_partition>::iterator;
+
+template<typename ResultIter, typename Predicate>
+partitions_request_iterator validate_replicas(
+  partitions_request_iterator begin,
+  partitions_request_iterator end,
+  ResultIter out_it,
+  error_code ec,
+  const ss::sstring& error_message,
+  Predicate&& p) {
+    auto valid_range_end = std::partition(
+      begin, end, std::forward<Predicate>(p));
+
+    std::transform(
+      valid_range_end,
+      end,
+      out_it,
+      [ec, &error_message](const reassignable_partition& partition) {
+          return reassignable_partition_response{
+            .partition_index = partition.partition_index,
+            .error_code = ec,
+            .error_message = error_message,
+          };
+      });
+    return valid_range_end;
+}
+
+template<typename ResultIter>
+partitions_request_iterator validate_partitions(
+  partitions_request_iterator begin,
+  partitions_request_iterator end,
+  ResultIter resp_it,
+  reassignable_topic_response topic_response,
+  std::vector<model::node_id> all_node_ids) {
+    // An undefined replicas vector is not an error, see "Replicas" in the
+    // AlterPartitionReassignmentsRequest schemata. Instead of checking for
+    // a null replicas vector in every call to validate_replicas, simply put
+    // those elements first and use iterator magic to run validation on
+    // defined replicas only.
+    auto valid_partitions_end = std::partition(
+      begin, end, [](const reassignable_partition& partition) {
+          return !partition.replicas.has_value();
+      });
+
+    std::vector<reassignable_partition_response> invalid_partitions;
+
+    valid_partitions_end = validate_replicas(
+      valid_partitions_end,
+      end,
+      std::back_inserter(invalid_partitions),
+      error_code::invalid_replica_assignment,
+      "Empty replica list specified in partition reassignment.",
+      [](const reassignable_partition& partition) {
+          return !partition.replicas->empty();
+      });
+
+    valid_partitions_end = validate_replicas(
+      valid_partitions_end,
+      end,
+      std::back_inserter(invalid_partitions),
+      error_code::invalid_replica_assignment,
+      "Duplicate replica ids in partition reassignment replica list",
+      [](const reassignable_partition& partition) {
+          auto replicas_copy = *partition.replicas;
+          std::sort(replicas_copy.begin(), replicas_copy.end());
+          auto no_duplicates_it = std::adjacent_find(
+            replicas_copy.cbegin(), replicas_copy.cend());
```

Comment:
you could insert here into an `absl::flat_hash_set` and look for entries that already exist.